From 15b3f686ce0d020a463a4b7784b0b3f0b949a099 Mon Sep 17 00:00:00 2001
From: Robert Dabrowski <robertx.dabrowski@intel.com>
Date: Thu, 19 Mar 2020 23:54:30 +0100
Subject: [PATCH] PONRTSYS-6953: Fix kernel crash seen for MIB counters update
 - change silent debug to error printout - fix sub_if range check - add lf to
 debug printouts

---
 .../net/datapath/mib_counters/dp_mib_counters.c    | 49 ++++++++++++----------
 1 file changed, 28 insertions(+), 21 deletions(-)

diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index d62ae5e2e390..574b3f2cad0c 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -178,13 +178,21 @@ s32 add_dev_to_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 {
 	struct mib_counters *dev_cnt;
 	struct pmac_port_info *port_info;
+	s32 sub_if_id_grp;
 
-	if ((!dev) ||
-	    (port_id < 0) || (port_id > MAX_PORT) ||
-	    (sub_if < 0) || (sub_if > MAX_SUBIF)) {
-		DP_DEBUG(DP_DBG_FLAG_MIB,
-			 "MIB CNT - net dev to add: name = %s, port_id = %d, sub_if = %d, inst = %d - WRONG DATA !\n",
-			 dev->name, port_id, sub_if, inst);
+	if ((!dev) || (port_id < 0) || (port_id > MAX_PORT)) {
+		pr_err("MIB CNT ERROR - WRONG data passed to %s: port_id %d, dev <%p>, dev->name %s\n",
+		       __func__, port_id, dev, dev ? dev->name : "[NULL]");
+		return -EINVAL;
+	}
+
+	port_info = get_dp_port_info(inst, port_id);
+	sub_if_id_grp = GET_VAP(sub_if, port_info->vap_offset,
+				port_info->vap_mask);
+
+	if ((sub_if_id_grp < 0) || (sub_if_id_grp > MAX_SUBIF)) {
+		pr_err("MIB CNT ERROR - WRONG data passed to %s: port_id %d, sub_if %d, dev %s - sub_if_id_grp %d !\n",
+		       __func__, port_id, sub_if, dev->name, sub_if_id_grp);
 		return -EINVAL;
 	}
 
@@ -194,13 +202,10 @@ s32 add_dev_to_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 	if (!dev_cnt)
 		return -ENOMEM;
 
-	port_info = get_dp_port_info(inst, port_id);
-
 	dev_cnt->inst = inst;
 	dev_cnt->dev = dev;
 	dev_cnt->port_id = port_id;
-	dev_cnt->sub_if_id_grp = GET_VAP(sub_if, port_info->vap_offset,
-					 port_info->vap_mask);
+	dev_cnt->sub_if_id_grp = sub_if_id_grp;
 	/* check if given net dev is on UNI side of bridge */
 	dev_cnt->dev_is_uni = (port_info->alloc_flags & DP_F_FAST_ETH_LAN) ||
 		(port_info->alloc_flags & DP_F_FAST_ETH_WAN);
@@ -231,24 +236,26 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 	struct list_head *p, *q;
 	s32 sub_if_id_grp;
 
-	if ((!dev) ||
-	    (port_id < 0) || (port_id > MAX_PORT) ||
-	    (sub_if < 0) || (sub_if > MAX_SUBIF)) {
-		DP_DEBUG(DP_DBG_FLAG_MIB,
-			 "MIB CNT - net dev to delete: name = %s, port_id = %d, sub_if = %d, inst = %d - WRONG DATA !\n",
-			 dev->name, port_id, sub_if, inst);
+	if ((!dev) || (port_id < 0) || (port_id > MAX_PORT)) {
+		pr_err("MIB CNT ERROR - WRONG data passed to %s: port_id %d, dev <%p>, dev->name %s\n",
+		       __func__, port_id, dev, dev ? dev->name : "[NULL]");
 		return -EINVAL;
 	}
 
 	port_info = get_dp_port_info(inst, port_id);
+	sub_if_id_grp = GET_VAP(sub_if, port_info->vap_offset,
+				port_info->vap_mask);
+
+	if ((sub_if_id_grp < 0) || (sub_if_id_grp > MAX_SUBIF)) {
+		pr_err("MIB CNT ERROR - WRONG data passed to %s: port_id %d, sub_if %d, dev %s - sub_if_id_grp %d !\n",
+		       __func__, port_id, sub_if, dev->name, sub_if_id_grp);
+		return -EINVAL;
+	}
 
 	DP_DEBUG(DP_DBG_FLAG_MIB,
 		 "MIB CNT - looking for item to delete: name = %s, port_id = %d, sub_if = %d, inst = %d\n",
 		 dev->name, port_id, sub_if, inst);
 
-	sub_if_id_grp = GET_VAP(sub_if, port_info->vap_offset,
-				port_info->vap_mask);
-
 	write_lock(&dev_list_lock);
 	list_for_each_safe(p, q, &registered_devices_list) {
 		mib_cnt = list_entry(p, struct mib_counters, dev_list);
@@ -258,7 +265,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 		    (mib_cnt->inst == inst) &&
 		    (!strcmp(mib_cnt->dev->name, dev->name))) {
 			DP_DEBUG(DP_DBG_FLAG_MIB,
-				 "MIB CNT - list item: name = %s, port_id = %d, sub_if_id_grp = %d, inst = %d - removed",
+				 "MIB CNT - list item: name = %s, port_id = %d, sub_if_id_grp = %d, inst = %d - removed\n",
 				 mib_cnt->dev->name,
 				 mib_cnt->port_id,
 				 mib_cnt->sub_if_id_grp,
@@ -267,7 +274,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 			devm_kfree(&g_dp_dev->dev, mib_cnt);
 		} else {
 			DP_DEBUG(DP_DBG_FLAG_MIB,
-				 "MIB CNT - list item: name = %s, port_id = %d, sub_if_id_grp = %d, inst = %d - doesn't match",
+				 "MIB CNT - list item: name = %s, port_id = %d, sub_if_id_grp = %d, inst = %d - doesn't match\n",
 				 mib_cnt->dev->name,
 				 mib_cnt->port_id,
 				 mib_cnt->sub_if_id_grp,
